### PR TITLE
Feature: One to many data structure

### LINKED ISSUE #285: Integrate FCОneToManyRelation

### PR DESCRIPTION
**Pull request by georgiyazovaliiski**
**Related issue(s)**:
Closes #285 

**Summary of the change**:
Removed invertible map, added 2 new data structures of type FCMap for TokenAssociations and AccountOwnership.
Unit tests are currently not working, due to the change of swirlds sdk to version 0.16.

**External impacts**:
None.

**Applicable documentation**
- [ ] Javadoc
- [ ] HAPI protobuf comments
- [ ] README

### CODE REVIEW

**Review by daniel-danev-limechain — APPROVED** [review]

File: hedera-node/src/main/java/com/hedera/services/context/ServicesContext.java [hunk 1]
```diff
@@ -570,7 +569,9 @@
 	private AtomicReference<FCMap<MerkleEntityId, MerkleSchedule>> queryableSchedules;
 	private AtomicReference<FCMap<MerkleBlobMeta, MerkleOptionalBlob>> queryableStorage;
 	private AtomicReference<FCMap<MerkleEntityAssociation, MerkleTokenRelStatus>> queryableTokenAssociations;
-	private AtomicReference<FCInvertibleHashMap<MerkleUniqueTokenId, MerkleUniqueToken, OwnerIdentifier>> queryableUniqueTokens;
+	private AtomicReference<FCMap<MerkleUniqueTokenId, MerkleUniqueToken>> queryableUniqueTokens;
+	private FCOneToManyRelation<EntityId, MerkleUniqueTokenId> uniqueTokenOwnership = new FCOneToManyRelation<>();
+	private FCOneToManyRelation<EntityId, MerkleUniqueTokenId> uniqueTokenAssociations = new FCOneToManyRelation<>();
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/context/ServicesContext.java:574]: We should place the new structures inside the `TypedTokenStore` and `rebuildViews` on `rebuildStoreViewsIfPresent`

File: hedera-node/src/main/java/com/hedera/services/context/ServicesContext.java [hunk 2]
```diff
@@ -2130,7 +2135,30 @@ public ExchangeRates midnightRates() {
...
+	public void rebuildOwnershipsAndAssociations() {
+		var uniqueTokens= state.uniqueTokens();
+		var keys = uniqueTokens.keySet();
+		for (MerkleUniqueTokenId key : keys) {
+			this.uniqueTokenAssociations.associate(key.tokenId(), key);
+			this.uniqueTokenOwnership.associate(uniqueTokens.get(key).getOwner(), key);
+		}
+	}
+
+	public FCOneToManyRelation<EntityId, MerkleUniqueTokenId> uniqueTokenAssociations() {
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/context/ServicesContext.java:2147]: We will have to have them as `AtomicReference<FCOneToManyRelation<EntityId, MerkleUniqueTokenId>> queriableTokenAssociations`

File: hedera-node/src/main/java/com/hedera/services/context/ServicesContext.java [hunk 3]
```diff
@@ -2130,7 +2135,30 @@ public ExchangeRates midnightRates() {
...
+		if (this.uniqueTokenAssociations == null) {
+			rebuildOwnershipsAndAssociations();
+		}
+		return this.uniqueTokenAssociations;
+	}
+
+	public FCOneToManyRelation<EntityId, MerkleUniqueTokenId> uniqueTokenAccountOwnerships() {
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/context/ServicesContext.java:2154]: We will have to have them as `AtomicReference<FCOneToManyRelation<EntityId, MerkleUniqueTokenId>> queriableTokenAccountOwnerships`

File: hedera-node/src/main/java/com/hedera/services/context/ServicesContext.java [hunk 4]
```diff
@@ -570,7 +569,9 @@
 	private AtomicReference<FCMap<MerkleEntityId, MerkleSchedule>> queryableSchedules;
 	private AtomicReference<FCMap<MerkleBlobMeta, MerkleOptionalBlob>> queryableStorage;
 	private AtomicReference<FCMap<MerkleEntityAssociation, MerkleTokenRelStatus>> queryableTokenAssociations;
-	private AtomicReference<FCInvertibleHashMap<MerkleUniqueTokenId, MerkleUniqueToken, OwnerIdentifier>> queryableUniqueTokens;
+	private AtomicReference<FCMap<MerkleUniqueTokenId, MerkleUniqueToken>> queryableUniqueTokens;
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/context/ServicesContext.java:572]: we will need to add 2 more AtomicReferences for `queryableTokenAssociations` and `queryableTokenAccountOwnerships`

File: hedera-node/src/main/java/com/hedera/services/context/ServicesContext.java [hunk 5]
```diff
@@ -2130,7 +2135,30 @@ public ExchangeRates midnightRates() {
 		return state.tokenAssociations();
 	}
 
-	public FCInvertibleHashMap<MerkleUniqueTokenId, MerkleUniqueToken, OwnerIdentifier> uniqueTokens() {
+	public void rebuildOwnershipsAndAssociations() {
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/context/ServicesContext.java:2138]: Lets do the rebuild inside the `TypedTokenStore`. We will need to create new instances of the FCOTMRs since we do not have clear.

File: hedera-node/src/main/java/com/hedera/services/records/TransactionRecordService.java [hunk 6]
```diff
@@ -136,6 +139,15 @@ public void includeOwnershipChanges(OwnershipTracker ownershipTracker) {
...
 								.setAccountNum(newOwner.getNum())).build());
+				var merkleUniqueTokenId = new MerkleUniqueTokenId(
+						EntityId.fromGrpcTokenId(tokenID),
+						change.getSerialNumber());
+				ownershipAssociation.disassociate(
+						new EntityId(previousOwner),
+						merkleUniqueTokenId);
+				ownershipAssociation.associate(
+						new EntityId(newOwner),
+						merkleUniqueTokenId);
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/records/TransactionRecordService.java:150]: This should be placed inside the `TypedToken` class -> `persistTrackers`

File: hedera-node/src/main/java/com/hedera/services/store/TypedTokenStore.java [hunk 7]
```diff
@@ -87,11 +88,15 @@ public TypedTokenStore(
 			AccountStore accountStore,
 			TransactionRecordService transactionRecordService,
 			Supplier<FCMap<MerkleEntityId, MerkleToken>> tokens,
-			Supplier<FCInvertibleHashMap<MerkleUniqueTokenId, MerkleUniqueToken, OwnerIdentifier>> uniqueTokens,
+			Supplier<FCMap<MerkleUniqueTokenId, MerkleUniqueToken>> uniqueTokens,
+			Supplier<FCOneToManyRelation<EntityId, MerkleUniqueTokenId>> uniqueOwnership,
+			Supplier<FCOneToManyRelation<EntityId, MerkleUniqueTokenId>> uniqueTokenAssociations,
 			Supplier<FCMap<MerkleEntityAssociation, MerkleTokenRelStatus>> tokenRels,
 			BackingTokenRels backingTokenRels
 	) {
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/TypedTokenStore.java:96]: In `persistToken` for minting you should not only do `uniqueTokens.get().put(merkleUniqueId, merkleUniqueToken);` but you should:
1) Add new `uniqueTokenAssociation` between the `Token` and the `newly created MerkleUniqueTokenId`
2) Add new `uniqueOwnership` between the `Treasury` and the `newly created MerkleUniqueTokenId`

Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/TypedTokenStore.java:96]: `loadUniqueToken` can be deleted

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]
You can delete the invertible hash map as-well

File: hedera-node/src/main/java/com/hedera/services/store/TypedTokenStore.java (hunk 7 above)
Comment (georgiyazovaliiski) [reply hedera-node/src/main/java/com/hedera/services/store/TypedTokenStore.java:103]: Don't we do this in the `persistTrackers` method? It is a generalized logic for all operations, when persisting `Change` objects to the new data structures.

File: hedera-node/src/main/java/com/hedera/services/records/TransactionRecordService.java [hunk 8]
```diff
@@ -21,6 +21,8 @@
  */
 
 import com.hedera.services.context.TransactionContext;
+import com.hedera.services.state.merkle.MerkleUniqueTokenId;
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/records/TransactionRecordService.java:24]: These imports are not necessary

File: hedera-node/src/main/java/com/hedera/services/store/TypedTokenStore.java [hunk 9]
```diff
@@ -271,24 +245,30 @@ public void persistToken(Token token) {
 		mapModelChangesToMutable(token, mutableToken);
 
 		if (token.hasMintedUniqueTokens()) {
-			final var currentNfts = nfts.get();
+			final var currentNfts = uniqueTokens.get();
+			final var currentUniqueTokenAssociations = uniqueTokenAssociations.get();
+			final var currentUniqueTokenAccountOwnerships = uniqueOwnership.get();
 			for (var uniqueToken : token.mintedUniqueTokens()) {
 				final var mintKey = new MerkleUniqueTokenId(
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/TypedTokenStore.java:252]: nit: lets rename `mintKey` to `merkleUniqueTokenId`. Someone might confuse `mintKey` for a cryptographic key. Same thing applies to `burnKey`

File: hedera-node/src/main/java/com/hedera/services/store/TypedTokenStore.java [hunk 10]
```diff
@@ -271,24 +245,30 @@ public void persistToken(Token token) {
 		mapModelChangesToMutable(token, mutableToken);
 
 		if (token.hasMintedUniqueTokens()) {
-			final var currentNfts = nfts.get();
+			final var currentNfts = uniqueTokens.get();
+			final var currentUniqueTokenAssociations = uniqueTokenAssociations.get();
+			final var currentUniqueTokenAccountOwnerships = uniqueOwnership.get();
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/TypedTokenStore.java:250]: I think that we can skip the variable declaration and use directly the `uniqueTokens.get();` method as-well as `uniqueTokenAssociations.get();` and `uniqueOwnership.get();`

Same thing for `burn`

File: hedera-node/src/main/java/com/hedera/services/store/TypedTokenStore.java [hunk 11]
```diff
@@ -271,24 +245,30 @@ public void persistToken(Token token) {
...
 			for (var uniqueToken : token.burnedUniqueTokens()) {
 				final var burnKey = new MerkleUniqueTokenId(
 						new EntityId(uniqueToken.getTokenId()), uniqueToken.getSerialNumber());
-				uniqueTokens.get().remove(burnKey);
 				currentNfts.remove(burnKey);
+				currentUniqueTokenAccountOwnerships.disassociate(new EntityId(uniqueToken.getOwner()), burnKey);
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/TypedTokenStore.java:270]: I think that we are not populating `owner` for the `uniqueToken` on burn. It would be best to directly use the `token.treasury` for the `new EntityId()` constructor. If we do that for `burn` it would be best to do it in `mint` for consistency

File: hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java [hunk 12]
```diff
@@ -152,6 +154,7 @@ public HederaTokenStore(
 		this.nftsLedger = nftsLedger;
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java:154]: I think that we no longer need `resolve` and `getUniqueToken` methods

File: hedera-node/src/main/java/com/hedera/services/txns/token/TokenMintTransitionLogic.java [hunk 13]
```diff
@@ -23,6 +23,7 @@
 import com.google.protobuf.ByteString;
 import com.hedera.services.context.TransactionContext;
 import com.hedera.services.state.enums.TokenType;
+import com.hedera.services.state.submerkle.RichInstant;
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/txns/token/TokenMintTransitionLogic.java:26]: not necessary change

File: hedera-node/src/test/java/com/hedera/services/ledger/BaseHederaLedgerTest.java [hunk 14]
```diff
@@ -99,6 +101,7 @@
 	protected AccountID rand = AccountID.newBuilder().setAccountNum(2_345).build();
 	protected AccountID deleted = AccountID.newBuilder().setAccountNum(3_456).build();
 	protected AccountID detached = AccountID.newBuilder().setAccountNum(4_567).build();
+	protected FCOneToManyRelation<EntityId, MerkleUniqueTokenId> uniqueTokenAccountOwnerships;
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/test/java/com/hedera/services/ledger/BaseHederaLedgerTest.java:104]: Is this change necessary?

File: test-clients/src/main/java/com/hedera/services/bdd/spec/fees/FeeCalculator.java [hunk 15]
```diff
@@ -134,7 +139,7 @@ public long forActivityBasedOpWithDetails(
 			AtomicReference<FeeObject> obs
 	) throws Throwable {
 		FeeData activityMetrics = metricsFor(txn, numPayerSigs, metricsCalculator);
-		return forOpWithDetails(op, SubType.DEFAULT, activityMetrics, obs);
+		return forOp(op, SubType.DEFAULT, activityMetrics);
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/fees/FeeCalculator.java:142]: This should not be changed. See:
https://github.com/hashgraph/hedera-services/blob/09d57eb17b71caa7de399a9c2045a8a413a5ee74/test-clients/src/main/java/com/hedera/services/bdd/spec/fees/FeeCalculator.java#L128

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]
InvertibleMap can be totally removed as-well

File: hedera-node/src/test/java/com/hedera/services/ledger/BaseHederaLedgerTest.java (hunk 14 above)
Comment (georgiyazovaliiski) [reply hedera-node/src/test/java/com/hedera/services/ledger/BaseHederaLedgerTest.java:104]: some unit tests were failing with null pointers

File: hedera-node/src/main/java/com/hedera/services/context/ServicesContext.java [hunk 16]
```diff
@@ -845,8 +846,10 @@ public IssEventInfo issEventInfo() {
 					() -> queryableTopics().get(),
 					() -> queryableAccounts().get(),
 					() -> queryableStorage().get(),
-					() -> queryableNfts().get(),
+					() -> queryableUniqueTokens().get(),
 					() -> queryableTokenAssociations().get(),
+					() -> queryableUniqueTokenAssociations().get(),
+					() -> queryableUniqueTokenAccountOwnerships().get(),
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/context/ServicesContext.java:852]: This part is not covered by unit tests

File: hedera-node/src/main/java/com/hedera/services/context/ServicesContext.java [hunk 17]
```diff
@@ -2070,20 +2077,27 @@ public Address address() {
...
+		}
+		return queryableUniqueTokenAssociations;
+	}
+
+	public AtomicReference<FCOneToManyRelation<EntityId, MerkleUniqueTokenId>> queryableUniqueTokenAccountOwnerships() {
+		if (queryableUniqueTokenAccountOwnerships == null) {
+			queryableUniqueTokenAccountOwnerships = new AtomicReference<>(uniqueTokenAccountOwnerships());
+		}
+		return queryableUniqueTokenAccountOwnerships;
+	}
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/context/ServicesContext.java:2099]: This is not covered by unit tests

File: hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java [hunk 18]
```diff
@@ -546,7 +567,7 @@ public boolean scheduleExists(ScheduleID id) {
 		return tokenAssociations;
 	}
 
-	void setNfts(Supplier<FCMap<MerkleUniqueTokenId, MerkleUniqueToken>> nfts) {
-		this.nfts = nfts;
+	void setNfts(Supplier<FCMap<MerkleUniqueTokenId, MerkleUniqueToken>> uniqueTokens) {
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java:570]: Is this method used somewhere?

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]

**Review by Daniel-K-Ivanov — APPROVED** [review]
